src/colorizer/recorders/CanvasRecorder.ts

```diff
@@ -0,0 +1,255 @@
+ * and cleanup.
+ */
+export default class CanvasRecorder {
+  private recording: boolean;
+  private finishedRecording: boolean;
+  private timerId: number;
+
+  protected setFrameAndRender: (frame: number) => Promise<void>;
+  protected getCanvas: () => HTMLCanvasElement;
+  protected options: RecordingOptions;
+
+  /**
+   * @param setFrameAndRender Async callback to set and update the currently displayed frame.
+   * @param getCanvas A callback to fetch the current canvas.
+   * @param options Configurable options for the recording.
+   */
+  constructor(
+    setFrameAndRender: (frame: number) => Promise<void>,
+    getCanvas: () => HTMLCanvasElement,
+    options?: Partial<RecordingOptions>
+  ) {
+    if (new.target === CanvasRecorder) {
+      throw new TypeError("CanvasRecorder is an abstract class. Do not construct instances of it directly!");
+    }
+
+    this.recording = false;
+    this.finishedRecording = false;
+    this.timerId = 0;
+
+    this.setFrameAndRender = setFrameAndRender;
+    this.getCanvas = getCanvas;
+
+    this.options = { ...defaultRecordingOptions, ...options };
+  }
+
+  /**
+   * Starts the recording process.
+   *
+   * Note that the recording will change the current frame and will not reset it once
+   * recording completes. Any cleanup should be done using the `onCompleted` callback parameter
+   * in the `recordingOptions` object or wherever `abort()` is called.
+   */
+  public async start(): Promise<void> {
+    if (this.recording || this.finishedRecording) {
+      return;
+    }
+    this.recording = true;
+
+    try {
+      await this.setup();
+    } catch (e) {
+      if (e instanceof Error) {
+        this.options.onError(e);
+      }
+      this.recording = false;
+      return;
+    }
+
+    this.startRecordingLoop();
+  }
+
+  private startRecordingLoop(): void {
+    // Reset any existing timers
+    clearTimeout(this.timerId);
+
+    const loadAndRecordFrame = async (frame: number): Promise<void> => {
+      if (!this.recording) {
+        return;
+      }
+      if (frame > this.options.max) {
+        this.recording = false;
+        await this.options.onCompleted();
+        return;
+      }
+
+      try {
+        await this.recordFrame(frame);
+      } catch (e) {
+        // Report errors but don't stop the recording.
+        if (e instanceof Error) {
+          this.options.onError(e);
+        } else if (e instanceof ErrorEvent) {
+          this.options.onError((e as ErrorEvent).error);
+        } else {
+          // May throw Event type if the error is from an event listener.
+          // This happens most often when a resource fails to load
+          // (e.g., network issues or user not on VPN)
+          // TODO: Update error message if this tool becomes public!
+          this.options.onError(
+            new Error(
+              "Encountered an unknown error while exporting. See the console for more details. For Institute users, please check VPN status."
```

Comment:
I like the todo.  This message seems like it's making big assumptions about who is reading it.  Might want to replace "the console" with "the browser's Javascript console"...